loader/src/main/java/net/neoforged/fml/common/asm/enumextension/RuntimeEnumExtender.java

```diff
@@ -0,0 +1,435 @@
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import net.neoforged.coremod.api.ASMAPI;
+import net.neoforged.fml.ModLoader;
+import net.neoforged.fml.ModLoadingIssue;
+import net.neoforged.fml.common.asm.ListGeneratorAdapter;
+import net.neoforged.neoforgespi.language.IModInfo;
+import org.jetbrains.annotations.ApiStatus;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+import org.objectweb.asm.commons.GeneratorAdapter;
+import org.objectweb.asm.commons.Method;
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.AnnotationNode;
+import org.objectweb.asm.tree.ClassNode;
+import org.objectweb.asm.tree.FieldInsnNode;
+import org.objectweb.asm.tree.FieldNode;
+import org.objectweb.asm.tree.InsnList;
+import org.objectweb.asm.tree.InsnNode;
+import org.objectweb.asm.tree.MethodInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+
+/**
+ * Transforms enums implementing {@link IExtensibleEnum} to add additional entries loaded from files provided by mods
+ */
+@ApiStatus.Internal
+public class RuntimeEnumExtender implements ILaunchPluginService {
+    private static final EnumSet<Phase> YAY = EnumSet.of(Phase.BEFORE);
+    private static final EnumSet<Phase> NAY = EnumSet.noneOf(Phase.class);
+    private static final Type MARKER_IFACE = Type.getType(IExtensibleEnum.class);
+    private static final Type INDEXED_ANNOTATION = Type.getType(IndexedEnum.class);
+    private static final Type NAMED_ANNOTATION = Type.getType(NamedEnum.class);
+    private static final Type RESERVED_ANNOTATION = Type.getType(ReservedConstructor.class);
+    private static final Type ENUM_PROXY = Type.getType(EnumProxy.class);
+    private static final Type NET_CHECK = Type.getType(NetworkedEnum.NetworkCheck.class);
+    private static final Type EXT_INFO = Type.getType(ExtensionInfo.class);
+    private static final String EXT_INFO_GETTER_DESC = Type.getMethodDescriptor(EXT_INFO);
+    private static final String EXT_INFO_CTOR_DESC = Type.getMethodDescriptor(
+            Type.VOID_TYPE, Type.BOOLEAN_TYPE, Type.INT_TYPE, Type.INT_TYPE, NET_CHECK);
+    private static final Type NETWORKED_ANNOTATION = Type.getType(NetworkedEnum.class);
+    private static final Type EXTENDER = Type.getType(RuntimeEnumExtender.class);
+    private static final Type ARRAYS = Type.getType("Ljava/util/Arrays;");
+    private static final int ENUM_FLAGS = Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC | Opcodes.ACC_FINAL | Opcodes.ACC_ENUM;
+    private static final int EXT_INFO_FLAGS = Opcodes.ACC_PRIVATE | Opcodes.ACC_STATIC | Opcodes.ACC_FINAL;
+    private static volatile Map<String, List<EnumPrototype>> prototypes = Map.of();
```

Comment:
`volatile` is not necessary since this happens in a single-threaded phase. Future uses of multi-threading should establish an appropriate `happens-before` relationship.